### PR TITLE
Add relative datetime queries for API

### PR DESCRIPTION
**Pull request by AlejandroMinaya**
## Summary
- Added `RelativeDateTimeField` that accept relative time expressions (e.g., `-1d`, `+2h`, `-30m`) and absolute time expressions (e.g. `2026-07-02 00:00:00`)
- Update filters to use this relative field for datetime values (excluding deprecated V1 filters)

🤖 Assisted-By: Claude Sonnet 4.5 <noreply@anthropic.com>

### CODE REVIEW

File: osidb/tests/test_relative_date_field.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+        field = RelativeDateField()
+
+        result = field.to_python("0d")
+
+        # 0 days should be today
+        expected_date = django_timezone.now().date()
+        assert result == expected_date
+
+    def test_leap_year_relative_values(self):
+        """Test that leap year calculations work (366 days ago)"""
```
Comment (osoukup) [inline osidb/tests/test_relative_date_field.py]: How do we know this or previous year was a leap one? I do not see any year pinning.

File: osidb/tests/test_relative_date_field.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        field = RelativeDateField()
+
+        # Test with 366 days (leap year scenario)
+        result = field.to_python("-366d")
+
+        expected_date = django_timezone.now().date() - timedelta(days=366)
+        assert result == expected_date
+
+    def test_extreme_past_value_exceeds_date_min(self):
+        """Test that values exceeding date.min raise OverflowError"""
```
Comment (osoukup) [inline osidb/tests/test_relative_date_field.py]: Is it necessary to care for some boundaries? I do not see any issue in letting the user to query for -5000d or whatever day want. Sure -5000y probably makes no sense but is there any technical reason not to allow it?

File: osidb/tests/test_relative_date_field.py [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+    def test_lowercase_m_is_minutes_for_date_field(self):
+        """-1m (minutes) when applied to date field"""
+        # Note: This tests edge case behavior - minutes don't make semantic sense
+        # for date-only fields, but we should handle it gracefully
+        field = RelativeDateField()
+        result = field.to_python("-1m")
+        # When relativedelta adds minutes to a date base, it returns datetime at 23:59 previous day
+        # For DateField, this is converted back to just the date part
+        assert isinstance(result, date)
+        assert result == date(2024, 6, 14)
```
[the remark this answers is not in the document]
Comment (AlejandroMinaya) [reply osidb/tests/test_relative_date_field.py]: Since our users may not be able to tell the difference between a DateTime and a Date field, it would create confusion if we raise an error for certain date params and not for others. The feedback is valid, but rejecting the request doesn't seem like a reasonable path forward

File: osidb/forms.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+from django.utils import timezone
+
+# Module-level constants for relative date/datetime parsing
+RELATIVE_TIME_UNIT_MAP = {
+    "s": "seconds",
+    "m": "minutes",
+    "h": "hours",
+    "d": "days",
+    "w": "weeks",
+}
```
Comment (osoukup) [inline osidb/forms.py]: Why no bigger period values?

File: osidb/forms.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        try:
+            return super().strptime(value, format)
+        except (ValueError, TypeError):
+            parsed = _parse_relative_value(value, timezone.now())
+            if parsed is not None:
+                return parsed
+            raise
+
+
+class RelativeDateField(forms.DateField):
```
Comment (osoukup) [inline osidb/forms.py]: I am not sure I understand the split between `Date` and `DateTime`, could you please explain why both are needed and where?

File: osidb/tests/test_relative_date_field.py [hunk 6]
```diff
@@ -338,6 +324,256 @@ def test_mixed_positive_and_negative_weeks(self):
         expected_pos = django_timezone.now().date() + timedelta(weeks=3)
         assert result_pos == expected_pos
 
+    # Month and year tests
+    @freeze_time("2024-06-15")
+    def test_one_month_ago(self):
+        """Test -1M returns date 1 month ago"""
+        field = RelativeDateField()
+        result = field.to_python("-1M")
```
Comment (osoukup) [inline osidb/tests/test_relative_date_field.py]: I saw somewhere that a moth is denoted by `mo` 🤔 Each has an advantage and disadvantage

**Review by osoukup — COMMENTED** [review]
I though that this is already ready for a review (was not a draft) but now I see I was probably incorrect and the PR is still being shaped. It seems to go well in the right direction.

File: osidb/forms.py (hunk 5 above)
Comment (AlejandroMinaya) [reply osidb/forms.py]: V1 Filters are using DateFilter instead of DateTimeFilter. In order to keep compatibility I created a RelativeDateFilter. I agree that it would be better to DateTimeFilter for everything, but I do not know if that change will break any underlying logic

File: osidb/tests/test_relative_date_field.py (hunk 6 above)
Comment (AlejandroMinaya) [reply osidb/tests/test_relative_date_field.py]: Yes! I saw it too. I am okay using `mo` but I do not know what would be better for our current UX. I can argue for and against either, but ultimately it will come down to what our users find more intuitive.

File: osidb/forms.py (hunk 4 above)
Comment (AlejandroMinaya) [reply osidb/forms.py]: Since supporting months and years meant considering more irregularities (28 vs. 30 vs. 31 days; leap years), I decided to first address the changes at this scale and figure out the support for larger periods later on. This is addressed in a later commit

File: osidb/tests/test_relative_date_field.py (hunk 6 above)
Comment (osoukup) [reply osidb/tests/test_relative_date_field.py]: I would probably maybe just discus that inside the DEV team and dictate one or the other way to the users :sweat_smile: not to make the process overly complicated. I can imagine that some would prefer one way and some the other but there would probably be no clear winner from the factual standpoint.

Comment (AlejandroMinaya) [reply osidb/tests/test_relative_date_field.py]: I am okay leaving 'M' until requested otherwise

Comment (osoukup) [reply osidb/tests/test_relative_date_field.py]: I probably do not have a strong preference here so you can just choose :wink:

File: osidb/tests/test_relative_datetime_filtering.py [hunk 7]
```diff
@@ -0,0 +1,858 @@
...
+        results = response.json()["results"]
+        uuids = {r["uuid"] for r in results}
+
+        assert str(affects["affect_5_hours_ago"].uuid) in uuids
+        assert str(affects["affect_2_hours_ago"].uuid) in uuids
+        assert str(affects["affect_30_min_ago"].uuid) not in uuids
+
+    # Note: The affects__created_dt filter uses the AffectV1 view which may not
+    # reflect timestamp changes made with auto_timestamps=False in tests.
+    # Commenting out this test as it requires investigation of the view refresh behavior.
```
[the remark this answers is not in the document]
Comment (AlejandroMinaya) [reply osidb/tests/test_relative_datetime_filtering.py]: I originally wanted to include this test for thoroughness but I wasn't aware at the moment that it depended on the AffectsV1 Materialized View. I think we can safely drop it since in the end it is using the same django-filter FilterSet and coupling this test with an AffectsV1 refresh is not worth it

File: osidb/tests/test_relative_date_field.py (hunk 1 above)
Comment (AlejandroMinaya) [reply osidb/tests/test_relative_date_field.py]: I still have to perform a cleaning sweep through the tests, but you're right to call this out. I don't think we need this test. Thank you

File: osidb/tests/test_relative_date_field.py (hunk 2 above)
Comment (AlejandroMinaya) [reply osidb/tests/test_relative_date_field.py]: I wanted to make sure that we handled correctly relative queries out of the date.min and date.max bounds but in either case it should behave the same as if you query for an absolute date outside this range. Do you think it is worth it to raise it as a ValidationError instead?

**Comment by AlejandroMinaya** [comment]
@coderabbitai full-review

**Comment by AlejandroMinaya** [comment]
@coderabbitai help

File: osidb/tests/test_relative_datetime_filtering.py [hunk 8]
```diff
@@ -0,0 +1,822 @@
...
+
+        response = auth_client().get(f"{test_api_uri}/flaws?created_dt__lte=-4M")
+        assert response.status_code == status.HTTP_200_OK
+        results = response.json()["results"]
+        uuids = {r["uuid"] for r in results}
+
+        assert str(flaws["flaw_6_months_ago"].uuid) in uuids
+        assert str(flaws["flaw_1_month_ago"].uuid) not in uuids
+        assert str(flaws["flaw_1_week_ago"].uuid) not in uuids
+
```
[the remark this answers is not in the document]
Comment (AlejandroMinaya) [reply osidb/tests/test_relative_datetime_filtering.py]: This is no longer necessary since this test was for V1 filters and these are deprecated

**Comment by AlejandroMinaya** [comment]
@coderabbitai full-review

File: osidb/tests/test_relative_datetime_filtering.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        assert response.status_code == status.HTTP_200_OK
+        results = response.json()["results"]
+        uuids = {r["uuid"] for r in results}
+
+        assert str(trackers["tracker_3_days_ago"].uuid) in uuids
+        assert str(trackers["tracker_1_day_ago"].uuid) in uuids
+        assert str(trackers["tracker_2_hours_ago"].uuid) not in uuids
+
+    # Note: affects__trackers__created_dt filtering has a bug in the filter configuration
+    # (uses method name instead of field name). Skipping this test until the filter is fixed.
```
[the remark this answers is not in the document]
Comment (AlejandroMinaya) [reply osidb/tests/test_relative_datetime_filtering.py]: V1 filters are deprecated

File: osidb/forms.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    sign, amount, unit = match.groups()
+    amount = int(amount) if sign != "-" else -int(amount)
+
+    # Case-sensitive lookup to distinguish 'm' (minutes) from 'M' (months)
+    delta_unit = RELATIVE_TIME_UNIT_MAP.get(unit)
+    if not delta_unit:
+        return None
+
+    delta_kwargs = {delta_unit: amount}
+    return base_value + relativedelta(**delta_kwargs)
```
[the remark this answers is not in the document]
Comment (AlejandroMinaya) [reply osidb/forms.py]: This has been addressed

File: osidb/tests/test_relative_datetime_field.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+    def test_parse_invalid_format_no_sign(self):
+        """Test that missing sign but valid number returns datetime (defaults to +)"""
+        # Note: Based on regex, this should actually work (sign is optional)
+        field = RelativeDateTimeField()
+
+        result = field.to_python("5h")
+        # The datetime should be approximately 1 day ago
+        expected_time = django_timezone.now() + timedelta(hours=5)
+
+        # Allow 1 second tolerance for test execution time
```
Comment (Elkasitu) [inline osidb/tests/test_relative_datetime_field.py]: For these sort of tests we like to use `freeze_time`, this makes comparing times directly easy and reliable.

File: osidb/filters.py [hunk 12]
```diff
@@ -81,6 +82,26 @@ class UUIDInFilter(BaseInFilter, UUIDFilter):
...
+    RelativeDateTimeFilter that accepts both absolute datetime values (default behavior)
+    and relative datetime strings like "-1d", "+2h", "-30m", etc.
+
+    Relative format: [+/-]<number><unit>
+    where unit is one of: s (seconds), m (minutes), h (hours), d (days), w (weeks)
+
+    Examples:
+        -1d     -> 1 day ago
+        +2h     -> 2 hours from now
+        -30m    -> 30 minutes ago
```
Comment (Elkasitu) [inline osidb/filters.py:100]: we should add an example without signs, the format just above hints at the possibility but it's not immediately clear, seems like we push for explicit sign usage

File: osidb/forms.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
+"""
+Custom form fields for OSIDB
```
Comment (Elkasitu) [inline osidb/forms.py]: Why? we don't use forms and the only place in which we use the logic herein is in `filters.py`, why not have this logic directly there?

File: osidb/filters.py [hunk 14]
```diff
@@ -791,6 +791,11 @@ class Meta:
             "references__url": ["exact"],
             "references__uuid": ["exact"],
         }
+        filter_overrides = {
+            models.DateTimeField: {
+                "filter_class": RelativeDateTimeFilter,
+            },
+        }
```
Comment (Elkasitu) [inline osidb/filters.py]: Instead of copy-pasting this for every `FilterSet`, we should either create a base class which each FilterSet inherits from or override it in `DJANGO_FILTERS.FILTERSET_BASE_CLASS`.

File: osidb/tests/test_relative_datetime_filtering.py [hunk 15]
```diff
@@ -0,0 +1,620 @@
...
+    # reflect timestamp changes made with auto_timestamps=False in tests.
+    # Commenting out this test as it requires investigation of the view refresh behavior.
+
+
+class TestRelativeDateTimeFilteringTrackers:
+    """
+    Test relative datetime filtering on Tracker endpoints
+    """
+
+    @pytest.fixture
```
Comment (Elkasitu) [inline osidb/tests/test_relative_datetime_filtering.py]: we tend to prefer fixtures in `conftest.py`

File: osidb/filters.py (hunk 14 above)
Comment (Elkasitu) [reply osidb/filters.py]: We might also want to replace explicit field declarations e.g. `created_dt = DateTimeFilter` for `created_dt = RelativeDateTimeFilter` as the `filter_overrides` only works for generated fields, it seems like you eventually realized and replaced `changed_after` and `changed_before` but we should do it for all explicitly declared `DateTimeFilter`

File: osidb/tests/test_relative_datetime_field.py [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+        assert abs(result - expected_time) < timedelta(seconds=1)
```
Comment (AlejandroMinaya) [inline osidb/tests/test_relative_datetime_field.py]: I'm dropping this test since it is a carry-over from initial iterations

File: osidb/tests/test_relative_datetime_field.py [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+
+    def test_leap_year_relative_values(self):
+        """Test that leap year calculations work (366 days ago)"""
+        field = RelativeDateTimeField()
+
+        # Test with 366 days (leap year scenario)
+        result = field.to_python("-366d")
+
+        expected_time = django_timezone.now() - timedelta(days=366)
+        assert abs(result - expected_time) < timedelta(seconds=1)
```
Comment (AlejandroMinaya) [inline osidb/tests/test_relative_datetime_field.py]: Dropping this test too

File: osidb/tests/test_relative_datetime_field.py [hunk 18]
```diff
@@ -145,30 +145,6 @@ def test_leap_year_relative_values(self):
...
-        field = RelativeDateTimeField()
-
-        # Calculate days that would exceed date.max (year 9999)
-        now = django_timezone.now()
-        days_to_max = (date.max - now.date()).days + 1000  # Go beyond max
-
-        # This should raise OverflowError when timedelta calculation exceeds range
-        with pytest.raises(OverflowError):
-            field.to_python(f"+{days_to_max}d")
-
```
Comment (Elkasitu) [inline osidb/tests/test_relative_datetime_field.py]: The commit message says that instead of `OverflowError` we raise `ValidationError` instead, this is not reflected in the tests

**Review by Elkasitu — CHANGES_REQUESTED** [review]
Heading in the right direction but requires some polish still, please ensure all commits are squashed into a single one before merging

**Review by osoukup — APPROVED** [review]
Looks well, good job!

**Comment by AlejandroMinaya** [comment]
In order to support OSIM Advanced Search I need to create a relative date field for DjangoQL. I will be pushing those changes later today

**Review by Elkasitu — APPROVED** [review]
